README
The Story of WvTFTPd
or
Why Would People Spend Time Writing a TFTP Server in this Day and Age?
======================================================================

WvTFTPd is a Trivial File Transfer Protocol server.

Why would anyone write a TFTP server in this day and age, you ask?  Well, TFTP
can be useful, but we at Net Integration Technologies thought that we could
make TFTP *much* more useful.  And using our WvStreams library, this task was
actually pretty simple.

Here's a summary of what makes WvTFTPd special.

- The synchronous request/acknowledgement protocol used by TFTP was
  improved by cheating slightly and using a technique we call "Negative
  Latency".  This is a little time consuming to explain, but basically
  it means "sending data before the other end asks for it".  In case
  cheating on protocols worries people (as it should), we have tested
  our server with a couple different TFTP clients, and we haven't 
  experienced any problems.  In addition, WvTFTPd ran 3.5 times faster 
  than the other servers out there!

- in the traditional TFTP servers, when a client requests a file, that exact
  pathname needs to exist: even if tftpd is restricted to only serve files from
  /tftpboot, it still requires the /tftpboot as part of the filename.  
  Our server doesn't mind if the client uses the full path or just a 
  filename (in which case the file is fetched from the default 
  directory).

- sometimes clients request the same file all the time, but you want to
  change which file goes with that client. For example, you might want
  to serve a different /tftpboot/kernel for different clients.  WvTFTPd 
  makes this very easy to configure.

- port management: traditionally, tftp daemons run from inetd even
  though they use udp, which needs strange workarounds (flushing buffers
  in odd places) and requires port number changes at bad times. ACK
  messages from the client go to a different port than the original
  request, which makes it hard to pass TFTP through a firewall.  WvTFTPd
  handles all connections on the standard TFTP port (69), so you only 
  need to forward that port to get TFTP access through a firewall.

- "adaptive retransmission": The TFTP protocol specifies that either the server
  or the client can timeout and resend requests.  This can cause some duplicate
  traffic, especially if a client isn't written correctly and responds to old
  DATA packets.  WvTFTP ignores all requests for retransmission and instead
  will retransmit after not receiving an ACK for two times the average return
  trip time so far.  Minimum and maximum timeouts are configurable.  This also
  applies to TFTP writes.

- large files: Some TFTP clients and servers can't handle files above 31 MB,
  due to the fact that the block number will roll over.  WvTFTP isn't one of
  them.  If you experience this problem, get a better client.

Compiling and Installing WvTFTPd
================================

You need to have a recent version of the WvStreams library and header files
installed on your system to use WvTFTPd.  We recommend version 4.0 or higher.
You can download WvStreams from

    http://open.nit.ca/wiki/index.php?page=WvStreams

You also require the "pkg-config" program, unless you want to modify the
Makefile and insert the appropriate paths.

WvTFTPd installs itself into <prefix>/sbin and the man page into
<prefix>/man, where <prefix> is the same prefix to which WvStreams is
installed (typically "/usr" or "/usr/local").  Again, edit the Makefile if
you wish to change this.  Eventually WvTFTPd might have a configure script
to do this.

To compile WvTFTPd, untar the package to somewhere like /usr/src and type
"make".  If there were no errors, type "make install".  Root privileges are,
of course, required to install the program.

Configuring WvTFTPd
===================

The configuration file for WvTFTPd is /etc/wvtftpd.conf.  WvTFTPd will 
run fine without any special configuration, but in order to take 
advantage of some of its special features, you'll need to create a 
configuration file.

The first section of the configuration file might look like this (default
values are shown):

[TFTP]
Base dir = /tftpboot/
Port = 69
Min Timeout = 100
Max Timeout = 5000
Max Timeout Count = 80
Prefetch = 3
Readonly = 1
Default File =
Strip Prefix = 
Overwrite existing file = 0
Client directory = 0
Create client directory = 0

"Base dir" is the default directory.  If a client requests a file 
without specifying the full path, the base dir is prepended.

"Port" specifies the port WvTFTP should use, if you don't want to use the
standard, 69, for some reason.

"Min Timeout" gives the minimum timeout value in milliseconds.  WvTFTP will
retransmit if it does not get an answer in twice the average RTT thus far or
the "Min Timeout" value, whichever is greater.  You can also specify a "Max
Timeout" as the maximum waiting time until retransmission.

If the number of timeouts reaches "Max Timeout Count", the transfer is
aborted.

"Prefetch" specifies the amount of negative latency, that is, how many
packets are sent out at a time.

"Readonly" determines if TFTP writes are allowed.  The default is 1 (writes
not allowed).

"Default File" is the file sent to a client if the requested file is 
unavailable.

The path given as "Strip Prefix" is automatically stripped from the
beginning of any client requests.  This is done before adding "base dir".

"Overwrite existing file" specifies if existing files should be overwritten
when the client uploads a file with the same name.  If the file exists and
this is set to 0 then the upload will fail.  The default is to not overwrite
existing files (0).

"Client directory" specifies if the client IP address should be appended to
the "Base dir" when a file is uploaded. This will cause clients to upload
into their own directory.  If the client directory does not exist and the
"Create client directory" directive is set to false (0) then the upload will
fail.  The default is to not append the client IP address to the "Base dir"
when a file is uploaded (0).

"Create client directory" will create the client's directory when the client
uploads a file (assuming "Client Directory" is set to 1; otherwise, this
option does nothing).  The directory is a subdirectory of "Base dir" and is
the IP address of the client (e.g /tftpboot/127.0.0.1). The default is to
not create the client directory (0).

The second section is [TFTP/Aliases].  It contains a list of filename
overrides.  You can specify an override for a particular IP address and also
default overrides for IPs not explicitly listed. Per-client filename
overrides look like "IPAddress/filename = clientnewfilename". Default
overrides look like "default/filename = newfilename".

For example:

[TFTP/Aliases]
default/image = image2_4.img
192.168.0.43/image = image2_5b.img

In this instance, if a user at 192.168.0.43 attempted to download the 
file "image", the file "image2_5b.img" would actually be sent.  Users 
from other machines will get the file "image2_4.img" when they request 
"image".

You can also specify one-time aliases in the [TFTP/Alias Once] section.  The
format is identical to [TFTP/Aliases]; you may have global or per-client
one-time aliases.  When a client asks for a file, WvTftp checks the
[TFTP/Alias Once] section first.  If a match is made, this alias is used,
and when the download finishes the alias is removed from the section
(regardless of whether the alias is global or client-specific).  Subsequent
matching requests will then be checked against [TFTP/Aliases] as normal. 
Note that the [TFTP/Alias Once] entry is only removed after a successful
download; the entry will be left alone if a download fails.

The last sections are [TFTP/Registered Clients] and [TFTP/New Clients]. 
[TFTP/Registered Clients] holds a list of client IP addresses ("192.168.0.43
= 1") that are known to the server.  When a client attempts to connect, if
its address is not in [TFTP/Registered Clients], it is added to
[TFTP/New Clients]. This has no function inside of WvTFTP itself but might
be useful in some situations (such as in our Net Integrators).

Note that UniConf, the configuration system that WvTFTPd uses, may rearrange
your config file such that all your settings, including [Aliases] and [New
Clients] and such, will be under the [TFTP] section.  Thus, your config may
end up looking like this:

[TFTP]
Port = 69
...
Aliases/default/image = image2_4.img
New Clients/192.168.0.1 = 1

This format and the above examples are completely interchangeable in
UniConf, but UniConf prefers the latter and will rearrange variables thusly. 
The above examples use separate sections just for clarity.


debian/changelog
wvtftpd (0.9.3) unstable; urgency=low

  * New upstream release.

 -- William Lachance <wlach@nit.ca>  Wed, 17 Aug 2005 17:35:00 -0500

wvtftpd (0.9.2) unstable; urgency=low

  * New upstream release.
  * Update build to require WvStreams 4.1 (unstable)

 -- William Lachance <wlach@nit.ca>  Wed, 06 Apr 2005 19:51:00 -0500

wvtftpd (0.10) unstable; urgency=low

  * New upstream release.
    - Fixes exploitable buffer overflow with privilege escalation.
  * Update build system to work with libwvstreams-dev 4.0.1.
  * Initial Debian release (Closes: #138949)

 -- Simon Law <sfllaw@debian.org>  Fri, 19 Nov 2004 00:39:59 -0500

wvtftpd (0.2) unstable; urgency=low

  * Synchronise with libwvstreams release
  * Obscure bug where wvtftpd would crash when the requesting entity was an
    etherboot host has been fixed.
  * Logs of bugs where wvtftpd would crash under load has been fixed.
  * Fixes to handle PXE Boot options

 -- Patrick Patterson <ppatters@debian.org>  Mon, 29 Jul 2002 18:39:51 -0400

wvtftpd (0.1) unstable; urgency=low

  * Initial Release.

 -- Patrick Patterson <ppatters@nit.ca>  Mon, 28 Jan 2002 13:04:52 -0500



wvtftpd.cc
/*
 * Worldvisions Weaver Software:
 *   Copyright (C) 1997-2005 Net Integration Technologies, Inc.
 *
 * This is the WvTFTP server daemon.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#include "uniconfroot.h"
#include "wvstreamsdaemon.h"
#include "wvtftpserver.h"
#include "wvver.h"
#include <signal.h>
#include <errno.h>

class WvTFTPDaemon : public WvStreamsDaemon
{
public:
    WvTFTPDaemon()
	: WvStreamsDaemon("wvtftpd", WVTFTP_VER_STRING, 
			  WvStreamsDaemonCallback(this, &WvTFTPDaemon::cb)),
          cfgmoniker("ini:/etc/wvtftpd.conf"), log("wvtftpd", WvLog::Info)
    {
	args.add_option('c', "config", "Config file",
			"ini:filename.ini", cfgmoniker);
    }
    
    virtual ~WvTFTPDaemon()
    {
	cfg.commit();
    }
    
    void cb(WvStreamsDaemon &daemon, void *)
    {
	cfg.unmount(cfg.whichmount(), true); // just in case
	cfg.mount(cfgmoniker);
	if (!cfg.whichmount() || !cfg.whichmount()->isok())
	{
	    log(WvLog::Error,
		"Can't read configuration from '%s'! Aborting.\n",
		cfgmoniker);
	    return;
	}
	
        tftps = new WvTFTPServer(cfg, 100);
        add_die_stream(tftps, true, "WvTFTP");
    }

private:
    WvString cfgmoniker;
    UniConfRoot cfg;
    WvTFTPServer *tftps;
    WvLog log; 
};


int main(int argc, char **argv)
{
    return WvTFTPDaemon().run(argc, argv);
}


wvtftpserver.cc
/*
 * Worldvisions Weaver Software:
 *   Copyright (C) 1997-2004 Net Integration Technologies, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "wvtftpserver.h"
#include "wvstrutils.h"
#include "wvtimeutils.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <ctype.h>
#include <unistd.h>

WvTFTPServer::WvTFTPServer(UniConf &_cfg, int _tftp_tick)
    : WvTFTPBase(_tftp_tick, _cfg["TFTP/Port"].getmeint(69)), cfg(_cfg)
{
    bool updated = update_cfg("TFTP Aliases", "TFTP/Aliases");
    updated |= update_cfg("TFTP Alias Once", "TFTP/Alias Once");
    if (updated)
	log(WvLog::Info, "Converted old-style TFTP configuration.\n");

    if (isok())
        log(WvLog::Info, "WvTFTP listening on %s.\n", *local());
    else
    {
        log(WvLog::Error, "Can't listen on port %s: %s\nAre you root?\n",
            cfg["TFTP/Port"].getmeint(69), errstr());
    }
}


WvTFTPServer::~WvTFTPServer()
{
    log(WvLog::Info, "WvTFTP shutting down.\n");
}


bool WvTFTPServer::update_cfg(WvStringParm oldsect, WvStringParm newsect)
{
    bool updated = false;
    UniConf::Iter i(cfg[oldsect]);
    for (i.rewind(); i.next(); )
    {
	if (i().haschildren())
	    continue;
	updated = true;
	WvString name(i().key().printable());
	char *p = strchr(name.edit(), ' ');
	if (!p)
	    cfg[newsect]["default"][name].setme(i().getme());
	else
	{
	    *p = 0;
	    cfg[newsect][name][p+1].setme(i().getme());
	}
	// i().remove(); // removing old entries kills downgrades!
    }
    cfg.commit();
    return updated;
}


void WvTFTPServer::execute()
{
    WvTFTPBase::execute();

    check_timeouts();

    if (!conns.isempty())
        alarm(tftp_tick);

    packetsize = read(packet, MAX_PACKET_SIZE);
    if (packetsize > 0)
    {
        dump_pkt();
        if (!conns[remaddr])
            new_connection();
        else
	{
	    TFTPOpcode opcode = (TFTPOpcode)(packet[0] * 256 + packet[1]);
	    
	    if (opcode == RRQ)
	    {
		// last transaction was interrupted, and they're starting over,
		// I guess.
		conns.remove(conns[remaddr]);
		new_connection();
	    }
	    else
		handle_packet();
	}
    }
}


void WvTFTPServer::check_timeouts()
{
    time_t timeout;

    TFTPConnDict::Iter i(conns);
    for (i.rewind(); i.next(); )
    {
        int expect_packet = (i->direction == tftpwrite) ? i->lastsent :
            i->unack;
        timeout = cfg["TFTP"]["Min Timeout"].getmeint(100);
        if (!i->total_packets)
            timeout = 1000;
        else if ((i->mult * i->mult * i->rtt / i->total_packets) > timeout)
            timeout = i->mult * i->mult * i->rtt / i->total_packets;
	
        struct timeval tv = wvtime();
        
        if (msecdiff(tv, *(i->pkttimes->get(expect_packet))) >= timeout)
        {
            i->numtimeouts++;

            log(WvLog::Debug1,
                "Timeout #%s (%s ms) on block %s from connection to %s.\n",
		i->numtimeouts, timeout, expect_packet, i->remote);
	    log(WvLog::Debug4, "[t1 %s, t2 %s, elapsed %s, expected %s]\n",
		tv.tv_sec, i->pkttimes->get(expect_packet)->tv_sec,
		msecdiff(tv, *(i->pkttimes->get(expect_packet))),
		expect_packet);
		
            log(WvLog::Debug4, "(packets %s, avg rtt %s, timeout %s, ms elapsed %s)\n",
                i->total_packets, 
                i->total_packets ? i->rtt / i->total_packets : 1000, timeout,
		msecdiff(tv, *(i->pkttimes->get(expect_packet))));
            if ((i->mult + 1) * (i->mult + 1) * i->rtt / i->total_packets <
		(time_t)cfg["TFTP"]["Max Timeout"].getmeint(5000))
                i->mult++;
            else
                log(WvLog::Debug1, "Max timeout reached.\n");
            log(WvLog::Debug4, "Increasing multiplier to %s.\n", i->mult * i->mult);

            // If the client times out too many times, stop sending it so much
            // at once - it could be choking on data.
            if ((i->numtimeouts % 5) == 0 && i->pktclump > 1) 
            {
                i->pktclump--;
                log(WvLog::Debug1, 
                        "Too many timeouts, reducing prefetch to %s\n", 
                        i->pktclump);
            }

            if (i->numtimeouts == cfg["TFTP/Max Timeout Count"].getmeint(80))
            {
                log(WvLog::Info,"Max timeouts reached; aborting transfer.\n");
                send_err(0, "Too many timeouts.");
                conns.remove(&i());
            }
            else if (i->send_oack)
            {
                log(WvLog::Debug4, "Sending oack ");
                memcpy(packet, i->oack, 512);
                packetsize = i->oacklen;
                dump_pkt();
                write(packet, packetsize);
                i->pkttimes->set(expect_packet, tv);
		i->timed_out_ignore = i->lastsent; 
            }
            else if (i->direction == tftpread)
	    {
                send_data(&i(), true);
		i->timed_out_ignore = i->lastsent; 
	    }
            else
	    {
                send_ack(&i(), true);
		i->timed_out_ignore = i->lastsent; 
	    }
	    
	    // 'i' might be invalid here!!
	    
	    // list might have lost an entry, screwing up the iterator;
	    // rewind it now.
	    i.rewind();
        }
    }
}


// Returns 0 if successful or the error number (1-8) if not.
int WvTFTPServer::validate_access(TFTPConn *c)
{
    WvString basedir = cfg["TFTP"]["Base dir"].getme("/tftpboot/");
    if (basedir[basedir.len() -1] != '/')
        basedir.append("/");

    if (strncmp(c->filename, basedir, basedir.len()))
        return 2;

    // As the original tftpd says, "prevent tricksters from getting
    // around the directory restrictions".
    if (!strncmp(c->filename, "../", 3))
        return 2;

    const char *cp;
    struct stat stbuf;
    for (cp = c->filename + 1; *cp; cp++)
        if(*cp == '.' && strncmp(cp-1, "/../", 4) == 0)
            return 2;
    if (stat(c->filename, &stbuf) < 0)
        if (c->direction == tftpread)
            return (errno == ENOENT ? 1 : 2);
        else
            return 0;            // If the file doesn't exist, allow writing.

    if (c->direction == tftpread)
    {
        if ((stbuf.st_mode&(S_IREAD >> 6)) == 0)
            return 2;
    }
    else
    {
        if ((stbuf.st_mode&(S_IWRITE >> 6)) == 0)
            return 2;
    }
    return 0; 
}


WvString WvTFTPServer::check_aliases(TFTPConn *c)
{
    if (!c)
	return WvString("");

    bool alias_once_fn_only = false;
    WvString clientportless(static_cast<WvIPAddr>(c->remote));

    WvString alias(cfg["TFTP/Alias Once"][clientportless]
		      [c->filename].getme(""));
    if (!alias)
    {
	alias = cfg["TFTP/Alias Once/default"][c->filename].getme("");
	alias_once_fn_only = true;
    }

    if (!!alias)
    {
	log(WvLog::Debug4, "Alias once is \"%s\".\n", alias);
	c->alias_once = true;
	c->alias = cfg["TFTP/Alias Once"]
	              [alias_once_fn_only ? WvString("default")
		                          : clientportless]
                      [c->filename];
    }
    else
    {
	alias = cfg["TFTP/Aliases"][clientportless][c->filename].getme(
	    cfg["TFTP/Aliases/default"][c->filename].getme("")
	    );
	log(WvLog::Debug4, "Alias is \"%s\".\n", alias);
    }

    return alias;
}


void WvTFTPServer::new_connection()
{
    log(WvLog::Info, "New connection from %s\n", remaddr);
    int code = packet[0]*256 + packet[1];
    log(WvLog::Debug4, "Packet opcode is %s.\n", code);
    if ((!code) || (code > 2))
    {
        log(WvLog::Debug, "Erroneous packet; aborting.\n");
        //send_err(4);
        return;
    }
    TFTPOpcode pktcode = static_cast<TFTPOpcode>(code);
    
    TFTPConn *c = new TFTPConn;
    c->remote = remaddr;
    WvIPAddr clientportless = static_cast<WvIPAddr>(c->remote);
    UniConfKey clientportlessk = UniConfKey(clientportless);

    if (!cfg["TFTP/Registered Clients"][clientportlessk].getmeint(
             cfg["TFTP/New Clients"][clientportlessk].getmeint(false)))
    {
        cfg["TFTP/New Clients"][clientportlessk].setmeint(true);
    }

    // Make sure the filename and mode actually end with nulls.
    bool foundnull1 = false;
    bool foundnull2 = false;
    unsigned int modestart = 0;
    for (unsigned int i = 2; i < packetsize; i++)
    {
        if (packet[i] == 0)
        {
            if (foundnull1)
            {
                foundnull2 = true;
                break;
            }
            else
            {
                foundnull1 = true;
                modestart = i + 1;
            }
        }
    }
    if (!foundnull1 || !foundnull2)
    {
        log(WvLog::Debug, "Badly formed packet; aborting.\n");
        send_err(4);
        delete c;
        return;
    }

    WvString origfilename = &packet[2];
    c->filename = origfilename;

    c->direction = static_cast<TFTPDir>((int)pktcode-1);
    log(WvLog::Debug4, "Direction is %s.\n", c->direction);
    if ((c->direction == tftpwrite) && cfg["TFTP"]["Readonly"].getmeint(1))
    {
        log(WvLog::Warning, "Writes are not permitted.\n");
        send_err(2);
        delete c;
        return;
    }

    // convert mode to TFTPMode type.
    unsigned int ch;
    for (ch = modestart; packet[ch] != 0; ch++)
        packet[ch] = tolower(packet[ch]);
    if (!strcmp(&packet[modestart], "netascii"))
        c->mode = netascii;
    else if (!strcmp(&packet[modestart], "octet"))
        c->mode = octet;
    else if (!strcmp(&packet[modestart], "mail"))
        c->mode = mail;
    else
    { 
        log(WvLog::Info, "Unknown mode string \"%s\"; aborting.\n",
            &packet[modestart]);
        send_err(4);
        delete c;
        return;
    }
    log(WvLog::Debug4, "Mode is %s.\n", c->mode);

    c->blksize = 512;
    c->tsize = 0;
    c->pktclump = cfg["TFTP"]["Prefetch"].getmeint(3);
    c->pkttimes = new PktTime(c->pktclump);
    c->unack = 0;
    c->donefile = false;
    c->numtimeouts = 0;
    c->rtt = 1000;
    c->total_packets = 1;
    c->mult = 1;
    c->timed_out_ignore = -1;
    
    for (int i = 0; i < c->pktclump; i++)
    {
        struct timeval tv = wvtime();
	c->pkttimes->set(i, tv);
    }

    if (c->direction == tftpread)
        log(WvLog::Info, "Client is requesting to read '%s'.\n",
	    origfilename);
    else
        log(WvLog::Info, "Client is requesting to write '%s'.\n",
	    origfilename);

    if (!check_filename(c))
    {
        delete c;
        return;
    }

    if (origfilename != c->filename)
	log(WvLog::Info, "...using '%s' instead.\n", c->filename);

    if (c->direction == tftpread)
    {
        if (c->mode == netascii)
            c->tftpfile = fopen(c->filename, "r");
        else if (c->mode == octet)
            c->tftpfile = fopen(c->filename, "rb");

        if (!c->tftpfile)
        {
            log(WvLog::Info, "Failed to open file for reading; aborting.\n");
            send_err(2);
            delete c;
            return;
        }
    }
    else
    {
        // check_filename() has already ensured that the file is not there
        // or that the user is allowed to overwrite it.
        umask(011);
        if (c->mode == netascii)
            c->tftpfile = fopen(c->filename, "w");
        else if (c->mode == octet)
            c->tftpfile = fopen(c->filename, "wb");

        if (!c->tftpfile)
        {
            log(WvLog::Info, "Failed to open file for writing; aborting.\n");
            send_err(3);
            delete c;
            return;
        }
    }

    c->send_oack = false;
    c->oacklen = 0;

    if (!process_options(c, ch+1))
    {
        delete c;
        return;
    }

    if (!strcmp(&packet[modestart], "netascii"))
      c->mode = netascii;
    else if (!strcmp(&packet[modestart], "octet"))
      c->mode = octet;
    else if (!strcmp(&packet[modestart], "mail"))
      c->mode = mail;
    else
    { 
	log(WvLog::Info, "Unknown mode string \"%s\"; aborting.\n",
	    &packet[modestart]);
	send_err(4);
	delete c;
	return;
    }

    alarm(tftp_tick);
    conns.add(c, true);
    if (c->direction == tftpread)
    {
        c->lastsent = 0;
        c->unack = 1;
        if (c->send_oack)
        {
            log(WvLog::Debug4, "Sending oack ");
            memcpy(packet, c->oack, 512);
            packetsize = c->oacklen;
            dump_pkt();
            write(packet, packetsize);
	    // Set pkttimes[1] to avoid timeouts on ACK for options.
	    struct timeval tv = wvtime();
	    c->pkttimes->set(1, tv);
        }
        else
        {
            log(WvLog::Debug4, "Last sent: %s unack: %s pktclump: %s\n",
                c->lastsent, c->unack, c->pktclump);
            int pktsremain = c->lastsent - c->unack;
            while (pktsremain < c->pktclump - 1)
            {
                log(WvLog::Debug4, "Result is %s\n", pktsremain);
                send_data(c);
                if (c->donefile)
                    break;
                pktsremain = c->lastsent - c->unack;
            }
        }
    }
    else
    {
        c->lastsent = -1;
        send_ack(c);
    }
}


bool WvTFTPServer::check_filename(TFTPConn *c)
{
    // Strip [TFTP]"Strip prefix" from filename, then compare it to 
    // [TFTP/Aliases] entries.  If nothing is found, then add [TFTP]"Base dir"
    // and try again.

    WvString strip_prefix = cfg["TFTP"]["Strip prefix"].getme("");
    if (strip_prefix != "")
    {
        if (strip_prefix[strip_prefix.len() -1] != '/')
            strip_prefix.append("/");

        log(WvLog::Debug4, "Strip prefix is %s.\n", strip_prefix);
        if (!strncmp(c->filename, strip_prefix, strip_prefix.len()))
        {
            log(WvLog::Debug4, "Stripping prefix.\n");
            c->filename = WvString(&c->filename[strip_prefix.len()]);
        }
    }
    log(WvLog::Debug4, "Filename after stripping is %s.\n", c->filename);

    WvString alias(check_aliases(c));

    if (!!alias)
        c->filename = alias;

    WvString basedir = cfg["TFTP"]["Base dir"].getme("/tftpboot/");
    if (basedir[basedir.len() -1] != '/')
        basedir.append("/");

    // Check to see if the clients write to their own directory and
    // the tftpdirection is write.  If so, then put the client IP address
    // bteween the basedir and the rest of the stuff.
    if (cfg["TFTP"]["Client directory"].getmeint() &&
        c->direction == tftpwrite)
    {
        basedir.append((const WvIPAddr) remaddr);
        basedir.append("/");
    }

    // If the first char isn't /, add base dir and look for aliases
    // again.  If we don't need to add the base dir, we've already
    // looked for aliases above.
    if (c->filename[0] != '/')
    {
	c->filename = WvString("%s%s", basedir, c->filename);
        if (alias == "")
        {
            // Check for aliases again
            log(WvLog::Debug4,
		"Filename before 2nd alias check is %s.\n", c->filename);
	    alias = check_aliases(c);
	    if (!!alias)
		c->filename = alias;
            log(WvLog::Debug4,
		"Filename after adding basedir and checking for alias is %s.\n",
                c->filename);
        }
    }

    if (c->direction == tftpwrite)
    {
        struct stat st;
        if (stat(c->filename, &st) == 0)
        {
            if (cfg["TFTP"]["Overwrite existing file"].getmeint())
            {
                if (!(st.st_mode & S_IWOTH))
                {
                    log(WvLog::Warning, "File is not world writable.\n");
                    send_err(2);
                    return false;
                }
                unlink(c->filename);
            }
            else
            {
                log(WvLog::Info, "File already exists; aborting.\n");
                send_err(6);
                return false;
            }
        }

        // Check to see if the directory exists for the client.  If not,
        // create it if the config is set.
        if (cfg["TFTP"]["Client directory"].getmeint())
        {
            log(WvLog::Debug, "BaseDir: %s\n", basedir);
            struct stat tftpdirstat;
            if (stat(basedir, &tftpdirstat) != 0)
            {
                if (errno == ENOENT &&
                    cfg["TFTP"]["Create client directory"].getmeint())
                {
                    // Create the directory.
                    if (mkdir(basedir, 0755))
                    {
                        // The directory creation failed.
                        log(WvLog::Warning, "Failed to create directory %s\n",
                            basedir);
                        send_err(2);
                        return false;
                    }
                }
                else
                {
                    // There was some kind of error other than the directory
                    // not existing.  Fail.
                    log(WvLog::Warning, "Client directory access failed: %s\n",
                        strerror(errno));
                    send_err(2);
                    return false;
                }
            }
            else if (!S_ISDIR(tftpdirstat.st_mode))
            {
                // There was some kind of error other than the directory
                // not existing (like the path is a file).  Fail.
                log(WvLog::Debug, "Specified path is not a directory: %s\n",
                    basedir);
                send_err(2);
                return false;
            }
        }
    }

    int tftpaccess = validate_access(c);
    if (tftpaccess == 1)
    {
        // File not found.  Check for default file.
        c->filename = cfg["TFTP"]["Default File"].getme("");
        if (!c->filename)
        {
            log(WvLog::Info, "File not found.  Aborting.\n");
            send_err(1);
	    return false;
        }
        alias = check_aliases(c);

        if (alias != "")
            c->filename = alias;
        if (c->filename[0] != '/')
            c->filename = WvString("%s%s", basedir, c->filename);

        tftpaccess = validate_access(c);
        if (tftpaccess)
        {
            log(WvLog::Warning, "File access failed (error %s).\n",
                tftpaccess);
            send_err(tftpaccess);
            return false;
        }
    }
    else if (tftpaccess)
    {
        log(WvLog::Warning, "File access failed (error %s).\n", tftpaccess);
        send_err(tftpaccess);
        return false;
    }
    log(WvLog::Debug4, "Filename is %s.\n", c->filename);
    return true;
}


unsigned int WvTFTPServer::process_options(TFTPConn *c, unsigned int opts_start)
{
    unsigned int p = opts_start;
    if (p < packetsize)
    {
        // One or more options available for reading.
        c->oack[0] = 0;
        c->oack[1] = 6;
        c->oacklen = 2;
        char *oackp = &(c->oack[2]);
        char *optname;
        char *optvalue = NULL; // Give optvalue a dummy value so gcc doesn't 
                               // complain about potential lack of
                               // initialization.
        while (p < packetsize-1)
        {
            optname = &packet[p];
            for (int i = 0; i < 2; i++)
            {
                while (packet[++p])
                {
                    if (p == packetsize)
                    {    
                        log(WvLog::Warning,
                            "Badly formed option %s.  Aborting.\n",
                            (i==0) ? "name" : "value");
                        send_err(8);
                        return 0;
                    }
                }
                if (!i)
                    optvalue = &packet[p+1];
            }
            p++;
            log(WvLog::Debug, "Option %s, value %s.\n", optname, optvalue);
            strlwr(optname);

            if (!strcmp(optname, "blksize"))
            {
                c->blksize = atoi(optvalue);
                if (c->blksize < 8 || c->blksize > 65464)
                {
                    WvString message("Request for blksize of %s is invalid.  "
                                     "Aborting.", c->blksize);
                    log(WvLog::Warning, "%s\n", message);
                    send_err(8, message);
                    return 0;
                }

		WvString oackblksize(c->blksize);
                log(WvLog::Debug, "Blksize option enabled (%s octets).\n", oackblksize);
                strcpy(oackp, optname);
                oackp += strlen(optname) + 1;
                c->oacklen += strlen(optname) + 1;
                strcpy(oackp, oackblksize);
		oackp += oackblksize.len() + 1;
		c->oacklen += oackblksize.len() + 1;
            }
            else if (!strcmp(optname, "timeout"))
                log(WvLog::Debug,
                    "Client request for timeout ignored.  Adaptive"
                    " retransmission is better.\n");
            else if (!strcmp(optname, "tsize"))
            {
                c->tsize = atoi(optvalue);
                if (c->tsize < 0)
                {
                    WvString message("Request for tsize of %s is invalid.  "
                                     "Aborting.", c->tsize);
                    log(WvLog::Warning, "%s\n", message);
                    send_err(8, message);
                    return 0;
                }

                if (c->tsize == 0 && c->direction == tftpread)
                {
                    struct stat tftpfilestat;
                    if (stat(c->filename, &tftpfilestat) != 0)
                    {
                        WvString message("Cannot get stats for file.  "
                                         "Aborting.");
                        log(WvLog::Warning, "%s\n", message);
                        send_err(8, message);
                        return 0;
                    }
                    c->tsize = tftpfilestat.st_size;
                }

                WvString oacktsize(c->tsize);
                log(WvLog::Debug, "Tsize option enabled (%s octets).\n", oacktsize);
                strcpy(oackp, optname);
                oackp += strlen(optname) + 1;
                c->oacklen += strlen(optname) + 1;
                strcpy(oackp, oacktsize);
                oackp += oacktsize.len() + 1;
                c->oacklen += oacktsize.len() + 1;
            }
        }
        if (c->oacklen)
            c->send_oack = true;
        else
            delete c->oack;
    }

    return p;
}
